File: python/large-class/good/user_service.py
class UserService:
    def __init__(self, db, email_config, payment_config):
        self.user_repository = UserRepository(db)
        self.email_service = EmailService(email_config)
        self.payment_service = PaymentService(payment_config)
        self.report_service = ReportService(db)
        self.activity_logger = ActivityLogger(db)

    # User management methods
    def create_user(self, email: str, name: str, password: str) -> int:
        hashed_password = Utility.hash_password(password)
        user_id = self.user_repository.create(email, name, hashed_password)

        self.email_service.send_welcome_email(email, name)
        self.activity_logger.log_activity(user_id, 'user_created')

        return user_id

    def authenticate_user(self, email: str, password: str):
        user = self.user_repository.authenticate(email, password)

        if user:
            self.activity_logger.log_activity(user.get_id(), 'user_login')

        return user

    def update_user_profile(self, user_id: int, name: str, email: str) -> None:
        user = self.user_repository.find_by_id(user_id)
        if user:
            user.update_profile(name, email)
            self.user_repository.update(user)
            self.activity_logger.log_activity(user_id, 'profile_updated')

    def get_user_balance(self, user_id: int) -> float:
        user = self.user_repository.find_by_id(user_id)
        return user.get_balance() if user else 0.0

    # Email methods
    def send_welcome_email(self, email: str, name: str) -> bool:
        return self.email_service.send_welcome_email(email, name)

    def send_password_reset_email(self, email: str, reset_token: str) -> bool:
        return self.email_service.send_password_reset_email(email, reset_token)

    def send_notification_email(self, email: str, subject: str, message: str) -> bool:
        return self.email_service.send_notification_email(email, subject, message)

    # Payment methods
    def process_stripe_payment(self, amount: float, token: str):
        return self.payment_service.process_stripe_payment(amount, token)

    def process_paypal_payment(self, amount: float, paypal_token: str):
        return self.payment_service.process_paypal_payment(amount, paypal_token)

    def refund_payment(self, transaction_id: str, amount: float):
        return self.payment_service.refund_payment(transaction_id, amount)

    # Reporting methods
    def generate_user_report(self, user_id: int):
        return self.report_service.generate_user_report(user_id)

    def generate_sales_report(self, start_date: str, end_date: str):
        return self.report_service.generate_sales_report(start_date, end_date)

File: golang/large-class/good/user_service.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

type UserRepository interface {
	Create(email, name, hashedPassword string) (int, error)
	Authenticate(email, password string) (*User, error)
	FindById(id int) (*User, error)
	Update(user *User) error
}

type EmailService interface {
	SendWelcomeEmail(email, name string) error
	SendPasswordResetEmail(email, resetToken string) error
	SendNotificationEmail(email, subject, message string) error
}

type PaymentService interface {
	ProcessStripePayment(amount float64, token string) map[string]interface{}
	ProcessPayPalPayment(amount float64, paypalToken string) map[string]interface{}
	RefundPayment(transactionId string, amount float64) map[string]interface{}
}

type ReportService interface {
	GenerateUserReport(userId int) map[string]interface{}
	GenerateSalesReport(startDate, endDate string) []map[string]interface{}
}

type ActivityLogger interface {
	LogActivity(userId int, action string) error
}

type UserService struct {
	userRepository UserRepository
	emailService   EmailService
	paymentService PaymentService
	reportService  ReportService
	activityLogger ActivityLogger
}

func NewUserService(
	userRepo UserRepository,
	emailSvc EmailService,
	paymentSvc PaymentService,
	reportSvc ReportService,
	activityLog ActivityLogger,
) *UserService {
	return &UserService{
		userRepository: userRepo,
		emailService:   emailSvc,
		paymentService: paymentSvc,
		reportService:  reportSvc,
		activityLogger: activityLog,
	}
}

// User management methods
func (us *UserService) CreateUser(email, name, password string) (int, error) {
	hashedPassword := hashPassword(password)
	userId, err := us.userRepository.Create(email, name, hashedPassword)
	if err != nil {
		return 0, err
	}

	us.emailService.SendWelcomeEmail(email, name)
	us.activityLogger.LogActivity(userId, "user_created")

	return userId, nil
}

func (us *UserService) AuthenticateUser(email, password string) (*User, error) {
	user, err := us.userRepository.Authenticate(email, password)
	if err != nil {
		return nil, err
	}

	if user != nil {
		us.activityLogger.LogActivity(user.GetId(), "user_login")
	}

	return user, nil
}

func (us *UserService) UpdateUserProfile(userId int, name, email string) error {
	user, err := us.userRepository.FindById(userId)
	if err != nil {
		return err
	}

	if user != nil {
		user.UpdateProfile(name, email)
		err = us.userRepository.Update(user)
		if err != nil {
			return err
		}
		us.activityLogger.LogActivity(userId, "profile_updated")
	}

	return nil
}

func (us *UserService) GetUserBalance(userId int) float64 {
	user, err := us.userRepository.FindById(userId)
	if err != nil {
		return 0.0
	}
	if user != nil {
		return user.GetBalance()
	}
	return 0.0
}

// Email methods - delegated to EmailService
func (us *UserService) SendWelcomeEmail(email, name string) error {
	return us.emailService.SendWelcomeEmail(email, name)
}

func (us *UserService) SendPasswordResetEmail(email, resetToken string) error {
	return us.emailService.SendPasswordResetEmail(email, resetToken)
}

func (us *UserService) SendNotificationEmail(email, subject, message string) error {
	return us.emailService.SendNotificationEmail(email, subject, message)
}

// Payment methods - delegated to PaymentService
func (us *UserService) ProcessStripePayment(amount float64, token string) map[string]interface{} {
	return us.paymentService.ProcessStripePayment(amount, token)
}

func (us *UserService) ProcessPayPalPayment(amount float64, paypalToken string) map[string]interface{} {
	return us.paymentService.ProcessPayPalPayment(amount, paypalToken)
}

func (us *UserService) RefundPayment(transactionId string, amount float64) map[string]interface{} {
	return us.paymentService.RefundPayment(transactionId, amount)
}

// Reporting methods - delegated to ReportService
func (us *UserService) GenerateUserReport(userId int) map[string]interface{} {
	return us.reportService.GenerateUserReport(userId)
}

func (us *UserService) GenerateSalesReport(startDate, endDate string) []map[string]interface{} {
	return us.reportService.GenerateSalesReport(startDate, endDate)
}

// Utility function
func hashPassword(password string) string {
	hash := sha256.Sum256([]byte(password))
	return hex.EncodeToString(hash[:])
}
